extension/src/helpers/hooks/useTokenDetails.tsx

```diff
@@ -0,0 +1,70 @@
+import { useReducer } from "react";
+import { useDispatch, useSelector } from "react-redux";
+
+import { tokenDetailsSelector, saveTokenDetails } from "popup/ducks/cache";
+import { getTokenDetails } from "@shared/api/internal";
+import { initialState, reducer } from "helpers/request";
+import { NetworkDetails } from "@shared/constants/stellar";
+import { AppDispatch, store } from "popup/App";
+
+export type TokenDetailsResponse = {
+  decimals: number;
+  symbol: string;
+  name: string;
+} | null;
+
+function useTokenDetails() {
+  const reduxDispatch = useDispatch<AppDispatch>();
+  const [state, dispatch] = useReducer(
+    reducer<TokenDetailsResponse, unknown>,
+    initialState,
+  );
+  const cachedTokenDetails = useSelector(tokenDetailsSelector);
+
+  const fetchData = async ({
+    contractId,
+    useCache = true,
+    publicKey,
+    networkDetails,
+  }: {
+    contractId: string;
+    useCache?: boolean;
+    publicKey: string;
+    networkDetails: NetworkDetails;
+  }): Promise<TokenDetailsResponse | Error> => {
+    dispatch({ type: "FETCH_DATA_START" });
+    try {
+      /*
+        Unlike the other cache hooks, this hook can be called multiple times within one render.
+        For example, when constructing the history rows, this hook can be called many times as we iterate over the history items. 
+        If we have cached token details earlier in the loop, we won't have access to the update redux state until the next render.
+        To workaround this, we will also check the redux state manually here rather than waiting for the next render to 
+        update the selector hook for us.
+      */
```

Comment:
Example: In my account history, I have 10 transfers using the same custom token. The getHistoryData hook will run once on first render and then call this method 10 times as we iterate over the history items. However, because we're still in the first render, the useSelector hook won't have access to the updated redux state yet. This will result in `cachedTokenDetails` being empty for all 10 iterations, resulting in 10 separate API calls to get the token details.